tensorflow/compiler/jit/xla_compilation_cache.cc

```diff
@@ -317,12 +321,112 @@ void LogOnceXlaCompiledFirstCluster() {
+    Entry* entry, const XlaCompiler::Options& options,
+    const std::vector<XlaCompiler::Argument>& args,
+    const string &function_name,
+    const std::function<Status(XlaCompiler* compiler,
+                               const std::vector<XlaCompiler::Argument>& args,
+                               XlaCompiler::CompilationResult*)>& compile_fn) {
+  tensorflow::Env* env = tensorflow::Env::Default();
+  const uint64 compile_start_us = env->NowMicros();
+
+  XlaCompiler compiler(options);
+  entry->compile_state = CompileState::kCompiled;
+
+  entry->compilation_status =
+      compile_fn(&compiler, args, &entry->compilation_result);
+  TF_RETURN_IF_ERROR(entry->compilation_status);
+  CHECK_EQ(entry->executable.get(), nullptr);
+  entry->compilation_status =
+      BuildExecutable(options, entry->compilation_result, &entry->executable);
+
+  const uint64 compile_end_us = env->NowMicros();
+  const uint64 compile_time_us = compile_end_us - compile_start_us;
+  metrics::UpdateXlaCompilationTime(compile_time_us);
+  {
+    mutex_lock lock(cluster_compile_stats_mu_);
+    auto it = cluster_compile_stats_.find(function_name);
+    const uint64 compile_time_s = compile_time_us / 1.0e6;
+    it->second.compile_count++;
+    it->second.cumulative_compile_time_us += compile_time_us;
+    it->second.max_compile_time_s = std::max(it->second.max_compile_time_s,
+                                             compile_time_s);
+    LogOnceXlaCompiledFirstCluster();
+    VLOG(1) << "compiled " << function_name << " "
+            << it->second.compile_count
+            << " times, compile time: " << compile_time_us
+            << " us, cumulative: " << it->second.cumulative_compile_time_us
+            << " us ("
+            << tensorflow::strings::HumanReadableElapsedTime(compile_time_s)
+            << " / "
+            << tensorflow::strings::HumanReadableElapsedTime(
+                   it->second.cumulative_compile_time_us / 1.0e6)
+            << ")";
+
+    XlaJitCompilationActivity jit_compilation_activity;
+    jit_compilation_activity.set_cluster_name(function_name);
+    jit_compilation_activity.set_compile_count(it->second.compile_count);
+    jit_compilation_activity.set_compile_time_us(compile_time_us);
+    jit_compilation_activity.set_cumulative_compile_time_us(
+        it->second.cumulative_compile_time_us);
+    TF_RETURN_IF_ERROR(
+        BroadcastXlaActivity(std::move(jit_compilation_activity)));
+  }
+
+  return Status::OK();
+}
+
+Status XlaCompilationCache::CompileAsynchronous(
+    Entry* entry, const XlaCompiler::Options& options,
+    const std::vector<XlaCompiler::Argument>& args,
+    const string &function_name,
+    const std::function<Status(XlaCompiler* compiler,
+                               const std::vector<XlaCompiler::Argument>& args,
+                               XlaCompiler::CompilationResult*)>& compile_fn) {
+  entry->compile_state = CompileState::kCompiling; // Still under caller's lock.
```

Comment:
Can you add a `CHECK` that verifies that we're still under the caller's lock?  Otherwise this could give us a nasty bug.

Reply:
no serious bug here. Worst case, compile_state is overwritten by another thread to kCompiled. This would happen anyway as part of the compilation. I will remove the comment, as this makes it look like a potential issue.